app/gui2/src/stores/project.ts

```diff
@@ -28,6 +31,57 @@ function resolveLsUrl(config: GuiConfig): LsUrls {
   throw new Error('Incomplete engine configuration')
 }
 
+async function initializeLsRpcConnection(urls: LsUrls): Promise<{
+  connection: LanguageServer
+  contentRoots: ContentRoot[]
+}> {
+  const transport = new WebSocketTransport(urls.rpcUrl)
+  const requestManager = new RequestManager([transport])
+  const client = new Client(requestManager)
+  const clientId = random.uuidv4() as Uuid
+  const connection = new LanguageServer(client)
+  const contentRoots = (await connection.initProtocolConnection(clientId)).contentRoots
+  return { connection, contentRoots }
+}
+
+/**
+ * Execution Context
+ *
+ * This class represent an execution context created in the Language Server. It creates
+ * it and pushes the initial frame upon construction.
+ *
+ * It hides the asynchronous nature of the language server. Each call is scheduled and
+ * run only when the previous call is done.
+ */
+export class ExecutionContext {
+  state: Promise<{ lsRpc: LanguageServer; id: ContextId }>
+
+  constructor(
+    lsRpc: Promise<LanguageServer>,
+    call: {
+      methodPointer: MethodPointer
+      thisArgumentExpression?: string
+      positionalArgumentsExpressions?: string[]
+    },
+  ) {
+    this.state = lsRpc.then(async (lsRpc) => {
+      const { contextId } = await lsRpc.createExecutionContext()
+      await lsRpc.pushExecutionContextItem(contextId, {
+        type: 'ExplicitCall',
+        positionalArgumentsExpressions: call.positionalArgumentsExpressions ?? [],
+        ...call,
+      })
+      return { lsRpc, id: contextId }
+    })
+  }
+
+  destroy() {
+    this.state.then(async ({ lsRpc, id }) => {
+      await lsRpc.destroyExecutionContext(id)
+    })
```

Comment:
this doesn't really have to be async/await.

```suggestion
    this.state.then(({ lsRpc, id }) => lsRpc.destroyExecutionContext(id))
```